File: DevExpress VCL/ExpressPrinting System/Demos/Delphi/PivotGridRL/PivotGridRLMain.pas
unit PivotGridRLMain;

interface

{$I ..\cxVer.inc}

uses
  Windows, Messages, SysUtils, Classes, Forms, Menus, Graphics, StdCtrls,
  dxPSCore, dxPSGlbl, dxPSUtl, cxControls, dxPScxCommon, ComCtrls, ShellApi,
{$IFDEF DELPHI6}
  Variants,
{$ENDIF}
  {dxPSPrVwAdv,} // enable for use of ExpressBars in PrintPreview
  ToolWin, Controls, ImgList, dxPSEngn, dxPrnPg, dxBkgnd, dxWrap, dxPrnDev,
  dxPSCompsProvider, dxPSFillPatterns, dxPSEdgePatterns, dxPScxPivotGrid2Lnk,
  cxCustomPivotGrid, cxPivotGrid, cxClasses, cxGraphics, cxCustomData,
  cxStyles, cxEdit;

type
  TPivotGridRLMainForm = class(TForm)
    MainMenu: TMainMenu;
    miFile: TMenuItem;
    miDesigner: TMenuItem;
    miHelp: TMenuItem;
    miPageSetup: TMenuItem;
    miPrintPreview: TMenuItem;
    miPrint: TMenuItem;
    N1: TMenuItem;
    miExit: TMenuItem;
    miDX: TMenuItem;
    dxComponentPrinter1: TdxComponentPrinter;
    ToolBar1: TToolBar;
    tbSetup: TToolButton;
    tbPreview: TToolButton;
    tbPrint: TToolButton;
    tbDevider: TToolButton;
    tbDesigner: TToolButton;
    tbDX: TToolButton;
    tbSeparator: TToolButton;
    tbExit: TToolButton;
    ilButtons: TImageList;
    PivotGrid: TcxPivotGrid;
    pgfPurchaseQuarter: TcxPivotGridField;
    pgfPurchaseMonth: TcxPivotGridField;
    pgfPaymentType: TcxPivotGridField;
    pgfQuantity: TcxPivotGridField;
    pgfCarName: TcxPivotGridField;
    pgfUnitPrice: TcxPivotGridField;
    pgfCompanyName: TcxPivotGridField;
    pgfPaymentAmount: TcxPivotGridField;
    dxComponentPrinter1Link1: TcxPivotGridReportLink;
    PaymentTypeImages: TImageList;
    procedure miDXClick(Sender: TObject);
    procedure miPageSetupClick(Sender: TObject);
    procedure miPrintPreviewClick(Sender: TObject);
    procedure miPrintClick(Sender: TObject);
    procedure miDesignerClick(Sender: TObject);
    procedure miExitClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure pgfPaymentTypeGetGroupImageIndex(Sender: TcxPivotGridField;
      const AItem: TcxPivotGridViewDataItem; var AImageIndex: Integer;
      var AImageAlignHorz: TAlignment;
      var AImageAlignVert: TcxAlignmentVert);
  private
  public
    { Public declarations }
  end;

var
  PivotGridRLMainForm: TPivotGridRLMainForm;

implementation


{$R *.DFM}

procedure TPivotGridRLMainForm.miDXClick(Sender: TObject);
begin
  ShellExecute(Handle, PChar('OPEN'), PChar('http://www.devexpress.com'), Nil, Nil, SW_SHOWMAXIMIZED);
end;

procedure TPivotGridRLMainForm.miPageSetupClick(Sender: TObject);
begin
  dxComponentPrinter1.PageSetup(nil);
end;

procedure TPivotGridRLMainForm.miPrintPreviewClick(Sender: TObject);
begin
  dxComponentPrinter1.Preview(True, nil);
end;

procedure TPivotGridRLMainForm.miPrintClick(Sender: TObject);
begin
  dxComponentPrinter1.Print(True, nil, nil);
end;

procedure TPivotGridRLMainForm.miDesignerClick(Sender: TObject);
begin
  dxComponentPrinter1.DesignReport(nil);
end;

procedure TPivotGridRLMainForm.miExitClick(Sender: TObject);
begin
  Close;
end;

procedure TPivotGridRLMainForm.FormCreate(Sender: TObject);
var
  AStream: TStream;
begin
  AStream := TResourceStream.Create(hInstance, 'PIVOTPREVIEWDATA', 'PIVOTDATA');
  try
    AStream.Position := 0;
    PivotGrid.DataController.LoadFromStream(AStream);
  finally
    AStream.Free;
  end;
end;

procedure TPivotGridRLMainForm.pgfPaymentTypeGetGroupImageIndex(
  Sender: TcxPivotGridField; const AItem: TcxPivotGridViewDataItem;
  var AImageIndex: Integer; var AImageAlignHorz: TAlignment;
  var AImageAlignVert: TcxAlignmentVert);
var
  Card: string;
begin
  Card := VarToStr(AItem.Value);
  if SameText(Card, 'AmEx') then
    AImageIndex := 0
  else
    if SameText(Card, 'Cash') then
      AImageIndex := 1
    else
      if SameText(Card, 'Master') then
        AImageIndex := 2
      else
        if SameText(Card, 'Visa') then
          AImageIndex := 3;
end;

end.

File: DevExpress VCL/ExpressPrinting System/Demos/CBuilder/PivotGridRL/PivotGridRLMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop
#include "PivotGridRLMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxControls"
#pragma link "cxCustomPivotGrid"
#pragma link "cxPivotGrid"
#pragma link "dxPSCore"
#pragma link "dxPScxCommon"
#pragma link "dxPScxPivotGrid2Lnk"
#pragma resource "*.dfm"
TPivotGridRLMainForm *PivotGridRLMainForm;
//---------------------------------------------------------------------------
__fastcall TPivotGridRLMainForm::TPivotGridRLMainForm(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TPivotGridRLMainForm::FormCreate(TObject *Sender)
{
  //#if __BORLANDC__ >= 0x610 // BCB >= 12
  #if defined(_DELPHI_STRING_UNICODE)
	PChar S = PChar(WideString("PIVOTDATA"));
  #else
	PChar S = "PIVOTDATA";
  #endif
  TStream * AStream = new TResourceStream((int)HInstance, "PIVOTPREVIEWDATA", S);
  try {
    AStream->Position = 0;
    PivotGrid->DataController->LoadFromStream(AStream);
   }
  __finally{
    delete AStream;
  }
}
//---------------------------------------------------------------------------

void __fastcall TPivotGridRLMainForm::miDXClick(TObject *Sender)
{
  ShellExecute(Handle, "OPEN", "http://www.devexpress.com", NULL, NULL, SW_SHOWMAXIMIZED);
}
//---------------------------------------------------------------------------

void __fastcall TPivotGridRLMainForm::miPageSetupClick(TObject *Sender)
{
  dxComponentPrinter1->PageSetup(NULL);
}
//---------------------------------------------------------------------------

void __fastcall TPivotGridRLMainForm::miPrintPreviewClick(TObject *Sender)
{
  dxComponentPrinter1->Preview(true, NULL);
}
//---------------------------------------------------------------------------

void __fastcall TPivotGridRLMainForm::miPrintClick(TObject *Sender)
{
  dxComponentPrinter1->Print(true, NULL, NULL);
}
//---------------------------------------------------------------------------

void __fastcall TPivotGridRLMainForm::miDesignerClick(TObject *Sender)
{
  dxComponentPrinter1->DesignReport(NULL);
}
//---------------------------------------------------------------------------

void __fastcall TPivotGridRLMainForm::miExitClick(TObject *Sender)
{
  Close();
}
//---------------------------------------------------------------------------


void __fastcall TPivotGridRLMainForm::pgfPaymentTypeGetGroupImageIndex(
      TcxPivotGridField *Sender, const TcxPivotGridViewDataItem *AItem,
      int &AImageIndex, TAlignment &AImageAlignHorz,
      TcxAlignmentVert &AImageAlignVert)
{
  AnsiString Card = ((TcxPivotGridViewDataItem *)AItem)->Value;
  if (SameText(Card, "AmEx")) AImageIndex = 0;
  else if (SameText(Card, "Cash")) AImageIndex = 1;
       else if (SameText(Card, "Master")) AImageIndex = 2;
            else if (SameText(Card, "Visa")) AImageIndex = 3;
}
//---------------------------------------------------------------------------

